### PR TITLE
feat: Add support for OR Queries Firestore

### PR DESCRIPTION
**Pull request by hemanshv**
feat: Add support for OR Queries Firestore

### CODE REVIEW

**Comment by Rishabh-V** [comment]
Looks like a few tests are failing. Can we please address them first? Also, please share the design doc with me (if it exists) so that I have the context. If not I'll sync up with you and get that information. Thanks.

**Comment by hemanshv** [comment]
@Rishabh-V just realized after raising the PR a small check was missing fixed it. yeah will share the doc.

Thanks

**Review by Rishabh-V — COMMENTED** [review]
I just had a look at the production code and I must admit that I don't have enough context to do a meaningful review. Let's catch up tomorrow and we can review it together in a meeting. Thanks.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore/Filter.cs
Comment (hemanshv) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore/Filter.cs:35]: To stop anyone else from trying to subclass it.

Comment (Rishabh-V) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore/Filter.cs:35]: Cool. Maybe put a documentation here.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore/Filter.cs
Comment (Rishabh-V) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore/Filter.cs:352]: This and below methods can use expression bodied members.

**Review by Rishabh-V — COMMENTED** [review]
Given my limited context and knowledge of Firestore, I suggest please request review directly from Jon after addressing/responding to the current comments. I 'll try to get involved from the starting from the next feature onwards.

**Comment by hemanshv** [comment]
Hi @jskeet ,
Please review this and share your valuable feedback.

Thanks,

**Review by jskeet — APPROVED** [review]
I'm basically fine with this, but there are various nits.

I've pushed more commits with some of the changes I've suggested that are hard to just describe. Please fetch those locally, and you can remove any commits that you don't want from there.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/QueryTest.cs
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/QueryTest.cs:993]: Why was this line removed? (It's much easier to spot now that there are fewer changes in this file.)

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/FilterTest.cs
Comment (hemanshv) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/FilterTest.cs:40]: yes Jon they are basically copied from Query test with some slight modification.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/QueryTest.cs
Comment (hemanshv) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/QueryTest.cs:993]: Earlier we used to convert object type to value type(serialization) right after it is passed into the `Where*` methods. Now we have delegated it to later (while converting to proto in `Filter.ToProto()`). So this statement will fail now.

Comment (jskeet) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/QueryTest.cs:993]: Hmm. That's a *bit* of a concern in terms of compatibility; I hadn't considered that before. For the moment, please highlight that in a separate test, as a record of the change. I'd like to think about that more carefully before we merge - it's just possible that users could be relying on the current behavior.

Comment (jskeet) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/QueryTest.cs:993]: I was considering that we *could* convert to a proto when checking for equality - but that also risks throwing an exception, if serialization fails. (Basically, changing the serialization timing knocks things somewhat.)

Another option to consider: make `Query.Filter` do the serialization. That's the earliest point at which we have all the required information - and it makes it fail early if serialization fails, as well as helping with equality operations. I haven't tried this though - I suspect it would add significant complexity.

Comment (jskeet) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/QueryTest.cs:993]: Next thought: serialize on a call to Query.Where(Filter), but discard the results - so we keep the same structure that's in the PR, but we're assured that serialization will work. We can then *also* serialize on GetHashCode and Equals, so we keep compatibility with previous code. It's slightly more expensive, but I don't expect filters to actually be that large.

**Comment by hemanshv** [comment]
Hi @jskeet , I've added the tests. Please have a look and let me know if anything else is needed.
Thanks,

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/QueryTest.cs
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/QueryTest.cs:1007]: Any reason this can't just be added to the unequal list above?

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/QueryTest.cs
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/QueryTest.cs:1003]: Nit: space after comma

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/FilterTest.cs
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/FilterTest.cs:53]: I'd split this up slightly, to make it clear where the exception is thrown:

```csharp
// No serialization check when the filter is created.
var filter = Filter.EqualTo("a.b", typeof(Filter));
// When used as part of a query, we check that it can be serialized in that context.
Assert.Throws<ArgumentException>(() => GetEmptyQuery().Where(filter);
```

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore/Query.cs
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore/Query.cs:49]: Nit: capital W.

(Having worked with codebases that don't use full sentences with appropriate punctuation and spacing for comments, I've learned that it's really distracting - at least for me. Hence nits that many other folks would be perfectly happy to let go...)

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/FilterTest.cs
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/FilterTest.cs:52]: I'd rename this to `Where_UnserializableFilter`. The "equal to" part isn't particularly interesting... it's that the filter, whatever it represents, can't be serialized.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/FilterTest.cs
Comment (jskeet) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/FilterTest.cs:51]: I haven't seen this test yet...

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/QueryTest.cs
Comment (hemanshv) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/QueryTest.cs:1007]: Here we want to test the equality when First Query(this) does not have a filter but 2nd query(other) have filters.

Comment (hemanshv) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/QueryTest.cs:1007]: Ok got it  Thanks for the explanation Jon

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/FilterTest.cs
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/FilterTest.cs:40]: Please indent this and line 37.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/FilterTest.cs
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/FilterTest.cs:301]: This is very hard to read at the moment - at least in GitHub. Please could you reformat it? I suspect if you break after `new[]` you'll be able to outdent by quite a lot. You could also consider creating the array elements in local variables before creating `expected`.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/QueryTest.cs
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/QueryTest.cs:1003]: Not sure why GitHub isn't letting me create a suggestion for this, but:

```csharp
col.Select("a") // Query without any filter (query._filters is null).
```

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/FilterTest.cs
Comment (jskeet) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/FilterTest.cs:54]: I'm still not seeing a test which combines the previous queries with filter.

Please write a test which starts with `var query = GetEmptyQuery.WhereEqualTo("a.b", "x").Where(Filter.EqualTo("a.c", "y"));` - that's not what NestedCompositeQuery tests.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/QueryTest.cs
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/QueryTest.cs:1003]: Sorry to keep nitting on this, but you've still got a lower-case q at the start...

**Review by jskeet — APPROVED** [review]
Just one final nit, then I think we're good to merge.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/QueryTest.cs
Comment (hemanshv) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/QueryTest.cs:1003]: Oh missed that will update, I am removing the `do not merge` label than

**Pull request merged**
